### ISSUE TITLE
Rest Gateway doesn't implement `completeAuthorize`

### ISSUE THREAD

**Issue by bummzack** [thephpleague/omnipay-paypal#170]
Since the Gateway implements `authorize` I expected to find a `completeAuthorize` method. 
How do you capture an authorized payment with the REST Gateway?

**Comment by lmatab** [comment]
did you find a way?

**Comment by bummzack** [comment]
I didn't look into it in detail, but I think the method _has_ to be implemented. Sadly I'm not very familiar with the PayPal API, so I can't create a PR for this.

**Comment by Zer0101** [comment]
I think this issue is redundant because REST gateway has [capture](https://github.com/thephpleague/omnipay-paypal/blob/82aa7c7548f688c051f7e354bc8b5df69193476b/src/RestGateway.php#L469), [void](https://github.com/thephpleague/omnipay-paypal/blob/82aa7c7548f688c051f7e354bc8b5df69193476b/src/RestGateway.php#L453) and [refund](https://github.com/thephpleague/omnipay-paypal/blob/82aa7c7548f688c051f7e354bc8b5df69193476b/src/RestGateway.php#L485) methods already.

**Comment by bummzack** [comment]
@Zer0101 Can you elaborate? How do you complete the authorization request then? Since `RestAuthorizeResponse` redirects you offsite, it should then call `completeAuthorize` once the authorization request has been completed offsite.

**Comment by Zer0101** [comment]
@bummzack, I think you must read doc-block for `capture` method first (see this [link](https://github.com/thephpleague/omnipay-paypal/blob/82aa7c7548f688c051f7e354bc8b5df69193476b/src/RestGateway.php#L469)).
If you still have doubts, then go to the PayPal documentation for authorization (see this [link](https://developer.paypal.com/docs/integration/direct/payments/authorize-and-capture-payments/)).
There you can read those lines:
```
Use the Payments API to authorize a payment that you can capture later. 
For example, you might have a delayed shipment for which you do not want to 
collect funds right away
```
as well the integration steps:
```
1. Meet the prerequisites.
2. Authorize the payment.
3. Get buyer approval and execute the payment authorization.
4. Capture the authorized payment.
```

There is no need for the `completeAuthorize` method, because authorization is `captured`, not `completed`.

**Comment by bummzack** [comment]
@Zer0101 I am aware of how authorize/capture works. But with an offsite gateway, you're supposed to issue an `authorize` request and then a `completeAuthorize` on your gateway once you come back from the offsite payment form. See: https://github.com/thephpleague/omnipay#gateway-methods

As you can see, the Restgateway implements `purchase` and [`completePurchase`](https://github.com/thephpleague/omnipay-paypal/blob/82aa7c7548f688c051f7e354bc8b5df69193476b/src/RestGateway.php#L414), therefore it has to implement `authorize` and `completeAuthorize` as well. 

You're confusing `completeAuthorize` with `capture`, which it is not. `authorize`, followed by `completeAuthorize` is required to create an authorized payment, which can then be captured later.

This is just how Omnipay handles things with offsite gateways and if a Gateway uses an offsite redirect, it should always implement both. `purchase` and `completePurchase` for a payment that gets settled immediately and `authorize` and `completeAuthorize` for an authorization that gets settled later.

**Comment by judgej** [comment]
@bummzack the gateway should only need to implement `completeAuthorize` if there is some data to capture or an action to perform on returning from the gateway. I'm not sure about this specific gateway, but there are afew that return the user with absolutely no context and require no further interaction with the gateway (the results being sent via a back-channel). Those gateways don't require Yo `completeAuthorize` support. It may be helpful to add it anyway, perhaps to fetch the transaction results from the gateway.

**Comment by bummzack** [comment]
@judgej thanks for chiming in. I wasn't aware of that rule. Good to know though, since I'm a maintainer of a library that integrates omnipay with a CMS and so far I've assumed that all gateways that redirect to an offsite form _have_ to implement `purchase` and `authorize` paired with their `complete` counterpart.
Apparently this isn't the case, so this issue can be closed.

**Comment by bummzack** [comment]
@judgej If I understand you right, `completePurchase` isn't mandatory either? So if I check `supportsCompletePurchase` and it returns `false`, I can mark a Payment directly as settled, without waiting for the completion? And similarly with `authorize`, if `supportsCompleteAuthorize` returns `false`, I can directly mark a payment as "Authorized" once it has gone through `authorize`?

**Comment by judgej** [comment]
Well, I'm not sure you can mark it as `authorized` just because there are no further actions to perform at the front end. Even some drivers that *do* support `completePurchase` and may indicate that the transaction was successful cannot be trusted, because the response from the gateway (brought back with the user's redirect) is not signed. Luckily, most are signed.

If there is no `completePurchase` support at all, then the success of the authorization comes from a different source, and that is normally a `notification` message that is captured at the back end and saved in the database.

However, it may be *helpful* in those instances if the `completePurchase` *was* implemented and just interrogated the gateway to get the result. So it may or may not be *required* for this particular gateway API (I've not checked that). And even if not required, it may still be *nice* to implement.

The main problem with consistency here, is that OmniPay does not have a standard or convention for joining what comes in from the back-channel `notification` to the application controller, so `completePurchase` has no standard way to ask your application, "did you get a notification that this was successful?". But `fetchTransaction` is widely used to interrogate the gateway to get the result, effectively bypassing the notification.

**Comment by Zer0101** [comment]
I checked it out and found, that `completeAuthorize` would make sense only in one case: if the REST gateway supports partial captures. Right now `is_final_capture` is hard-coded to `true` in `RestCaptureRequest`, so `capture` acts as  `completeAuthorize`. 
Otherwise it would be like this: 
```
    public function capture(array $parameters = array())
    {
        return $this->createRequest('\Omnipay\PayPal\Message\RestCaptureRequest', $parameters);
    }
```
```
    public function completeAuthorize(array $parameters = array())
    {
        return $this->createRequest('\Omnipay\PayPal\Message\RestCaptureRequest', $parameters + ['isFinalCapture' => true]);
    }
```

**Comment by bummzack** [comment]
@judgej Thanks for the explanation and sorry if this discussion doesn't really belong here. In the omnipay wrapper I'm maintaining, we already have support for notifications and we mark payments accordingly with a `pending` state. I did the same for offsite gateways (eg. marking as `pending` until `completePurchase` was called), but this assumption was wrong.

Developers that use our library can supply a config-flag if a payment should wait for an offsite notification and also implement custom handlers to deal with the incoming notification. I just want to provide a sensible default and not break gateways like this one with our assumptions how the process has to go.

So to summarize: If the gateway doesn't get updates via a notification and it doesn't implement the `completeAuthorize` method, it's safe to assume that `authorize` alone has been sufficient to authorize a payment.

@Zer0101 I think you're still confusing `capture` and `completeAuthorize`. `completeAuthorize` will **not** capture anything.

**Comment by rebelpixel** [comment]
Very late to this thread. I think `completeAuthorize` and `completePurchase` methods are made unnecessary because PayPal's Rest Gateway uses Instant Payment Notifications (IPN) that basically pings your site and sends the necessary transaction information via that method. Thus, by the time you are redirected after the transaction at PayPal's servers, you should've received the details of the transaction and tailor your redirected confirmation page accordingly.

---

2023-02-01 EDIT: I've built something recently using `omnipay-paypal`, and my comment above may not be the exact solution to this. The current version for `PayPal_Rest` implements `completePurchase` and this method works the same way as `completeAuthorize` as they essentially just execute the action. This can be verified by looking at the docs of PayPal's API v1, as well as what is shown in their [SDK sample](https://github.com/paypal/PayPal-PHP-SDK/blob/1a2ed767bb09374a8a222069930e94fccf99009e/sample/payments/AuthorizePaymentUsingPayPal.php), with the action passed to the same [returnUrl](https://github.com/paypal/PayPal-PHP-SDK/blob/1a2ed767bb09374a8a222069930e94fccf99009e/sample/payments/ExecutePayment.php) after being processed off-site.
